billing-report-agent/src/main/java/com/epam/pipeline/billingreportagent/service/impl/converter/RunToBillingRequestConverter.java

```diff
@@ -0,0 +1,172 @@
+     * @param syncStart time point, where the whole synchronization process was started
+     * @return list of requests to be performed (deletion index request if no billing requests created)
+     */
+    @Override
+    public List<DocWriteRequest> convertEntityToRequests(final EntityContainer<PipelineRun> container,
+                                                         final String indexName,
+                                                         final LocalDateTime syncStart) {
+            return convertRunToBillings(container, syncStart).stream()
+                .map(billingInfo -> getDocWriteRequest(indexName, container, billingInfo))
+                .collect(Collectors.toList());
+    }
+
+    protected List<PipelineRunBillingInfo> convertRunToBillings(final EntityContainer<PipelineRun> run,
+                                                                final LocalDateTime syncStart) {
+        final BigDecimal pricePerHour = run.getEntity().getPricePerHour();
+        final List<RunStatus> statuses = run.getEntity().getRunStatuses();
+        adjustStatuses(statuses, syncStart, run.getEntity().getId());
+        final List<List<RunStatus>> activePeriods = calculateActivityPeriods(statuses);
+
+        final Map<LocalDate, PipelineRunBillingInfo> reports = new HashMap<>();
+        activePeriods.forEach(period -> {
+            for (int i = 0; i < period.size() - 1; i++) {
+                final LocalDateTime start = period.get(i).getTimestamp();
+                final LocalDateTime end = period.get(i + 1).getTimestamp();
+                final Long dailyCost = calculateCostsForPeriod(start, end, pricePerHour);
+                addDailyCostReport(reports, start.toLocalDate(), dailyCost, ChronoUnit.MINUTES.between(start, end), run);
+            }
+        });
+        return new ArrayList<>(reports.values());
+    }
+
+    private void adjustStatuses(final List<RunStatus> statuses, final LocalDateTime syncStart, final Long runId) {
+        statuses.sort(Comparator.comparing(RunStatus::getTimestamp));
+        final RunStatus lastStatus = statuses.get(statuses.size() - 1);
+        if (lastStatus.getStatus() == TaskStatus.RUNNING) {
+            statuses.add(new RunStatus(null, null, syncStart));
+        } else if (statuses.size() > 1) {
+            final RunStatus status = statuses.get(statuses.size() - 2);
+            if (status.getStatus() != TaskStatus.RUNNING) {
+                statuses.remove(statuses.size() - 1);
+            }
+        }
+        if (statuses.size() % 2 != 0) { // no RUNNING states for given run
+            log.warn("Can't adjusting run {} statuses, it has no RUNNING states in history!", runId);
+            statuses.clear();
+        }
+    }
+
+    private List<List<RunStatus>> calculateActivityPeriods(final List<RunStatus> statuses) {
+        final List<List<RunStatus>> activePeriods = new ArrayList<>();
+        for (int i = 0; i < statuses.size(); i += 2) {
```

Comment:
I'm not sure about this approach. As far as I understand, it fully relies that we'll have a pair of statuses `active` and `passive`. Maybe we can just iterate over statuses and calculate active period when status changes from `active` to `passive`?

Reply:
Done, iterating over statuses, searching change from active to passive and calculating cost for that period